File: others/1600-1700/1649. 通过指令创建有序数组.py
from typing import List

MOD = int(1e9 + 7)


class FenwickTree:
    def __init__(self, n: int, mx: int) -> None:
        self.tree = [0] * (n + 5)
        self.mx = mx

    def update(self, index: int, val: int) -> None:
        i = index + 1
        while i < len(self.tree):
            self.tree[i] += val
            i += i & -i

    def pre(self, index: int) -> int:
        s = 0
        while index > 0:
            s += self.tree[index]
            index &= index - 1
        return s

    def queryGreater(self, val: int) -> int:
        return self.pre(self.mx + 1) - self.pre(val + 1)

    def queryLesser(self, val: int) -> int:
        return self.pre(val)


class Solution:
    def createSortedArray(self, instructions: List[int]) -> int:
        ft = FenwickTree(max(instructions), max(instructions))
        ans = 0

        for i, x in enumerate(instructions):
            ans = (ans + min(ft.queryGreater(x), ft.queryLesser(x))) % MOD
            ft.update(x, 1)
        return ans

File: leetcode/1600-1700/1649. 通过指令创建有序数组.cpp
#include <bits/stdc++.h>
using namespace std;
vector<vector<int>> DIRS = {{1, 1}, {1, -1}, {-1, -1}, {-1, 1}};
using ll = long long;
constexpr int MOD = int(1e9 + 7);

class FenwickTree {
private:
    vector<int> tree;

public:
    FenwickTree(int n) : tree(n + 1) {}

    void update(int index, int val = 1) {
        int i = index;
        for (; i < tree.size(); i += i & -i) {
            tree[i] += val;
        }
    }

    int pre(int index) {
        int s = 0;
        while (index > 0) {
            s += tree[index];
            index -= index & -index;
        }
        return s;
    }
};


class Solution {
public:
    int createSortedArray(vector<int> &instructions) {
        int n = instructions.size();
        ll ans = 0;
        int mx = ranges::max(instructions);

        FenwickTree ft(mx);
        for (int x: instructions) {
            int less = ft.pre(x - 1);
            int greater = ft.pre(mx) - ft.pre(x);
            ans = (ans + min(less, greater)) % MOD;
            ft.update(x);
        }
        return ans;
    }
};
